src/proto/streams/flow_control.rs

```diff
@@ -1,121 +1,86 @@
+        if self.window_size < 0 {
+            return false;
         }
 
-        plus - self.underflow
+        self.window_size as WindowSize > self.available
     }
 
-    /// Returns true iff `claim_window(sz)` would return succeed.
-    pub fn ensure_window<T>(&mut self, sz: WindowSize, err: T) -> Result<(), ConnectionError>
-        where T: Into<ConnectionError>,
-    {
-        if sz <= self.window_size {
-            Ok(())
-        } else {
-            Err(err.into())
-        }
+    pub fn claim_capacity(&mut self, capacity: WindowSize) {
+        assert!(self.available >= capacity);
+        self.available -= capacity;
     }
 
-    /// Reduce future capacity of the window.
-    ///
-    /// This accomodates updates to SETTINGS_INITIAL_WINDOW_SIZE.
-    pub fn shrink_window(&mut self, dec: WindowSize) {
-        /*
-        if decr < self.next_window_update {
-            self.next_window_update -= decr
-        } else {
-            self.underflow += decr - self.next_window_update;
-            self.next_window_update = 0;
-        }
-        */
+    pub fn assign_capacity(&mut self, capacity: WindowSize) {
+        assert!(self.window_size() >= self.available + capacity);
+        self.available += capacity;
     }
 
-
-    /// Claims the provided amount from the window, if there is enough space.
+    /// Update the window size.
     ///
-    /// Fails when `apply_window_update()` hasn't returned at least `sz` more bytes than
-    /// have been previously claimed.
-    pub fn claim_window<T>(&mut self, sz: WindowSize, err: T)
-        -> Result<(), ConnectionError>
-        where T: Into<ConnectionError>,
-    {
-        self.ensure_window(sz, err)?;
-
-        self.window_size -= sz;
-        Ok(())
-    }
-
-    /// Increase the _unadvertised_ window capacity.
-    pub fn expand_window(&mut self, sz: WindowSize)
-        -> Result<(), ConnectionError>
-    {
+    /// This is called after receiving a WINDOW_UPDATE frame
+    pub fn inc_window(&mut self, sz: WindowSize) -> Result<(), ConnectionError> {
         // TODO: Handle invalid increment
-        if sz <= self.underflow {
-            self.underflow -= sz;
-            return Ok(());
-        }
-
-        let added = sz - self.underflow;
-        self.next_window_update += added;
-        self.underflow = 0;
-
+        self.window_size += sz as i32;
         Ok(())
     }
 
-    /*
-    /// Obtains the unadvertised window update.
-    ///
-    /// This does not apply the window update to `self`.
-    pub fn peek_window_update(&mut self) -> Option<WindowSize> {
-        if self.next_window_update == 0 {
-            None
-        } else {
-            Some(self.next_window_update)
-        }
+    /// Decrements the window reflecting data has actually been sent. The caller
+    /// must ensure that the window has capacity.
+    pub fn send_data(&mut self, sz: WindowSize) {
+        assert!(sz <= self.window_size as WindowSize);
+        self.window_size -= sz as i32;
```

Comment:
as noted in chat, we'll also need to account for what happens when the initial window size is decreased.